### PR TITLE
RYA-392-Datetime-Within

### PR DESCRIPTION
**Pull request by meiercaleb**
## Description
Added a Function DateTimeWithinPeriod to determine whether two datetimes were within a
specified period of time of one another.

### Tests
Added unit tests to test the Function and added integration tests to verify that the Function
is correctly utilized during query evaluation.

### Links
[Jira](https://issues.apache.org/jira/browse/RYA-392)

### Checklist
- [ ] Code Review
- [x] Squash Commits

#### People To Reivew
@jdasch @kchilton2

### CODE REVIEW

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/incubator-rya-master-with-optionals-pull-requests/512/

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/incubator-rya-master-with-optionals-pull-requests/513/<h2>Failed Tests: <span class='status-failure'>1</span></h2><h3><a name='incubator-rya-master-with-optionals-pull-requests/org.apache.rya:rya.indexing' /><a href='https://builds.apache.org/job/incubator-rya-master-with-optionals-pull-requests/513/org.apache.rya$rya.indexing/testReport'>incubator-rya-master-with-optionals-pull-requests/org.apache.rya:rya.indexing</a>: <span class='status-failure'>1</span></h3><ul><li><a href='https://builds.apache.org/job/incubator-rya-master-with-optionals-pull-requests/513/org.apache.rya$rya.indexing/testReport/org.apache.rya.indexing.mongo/MongoEntityIndexTest/org_apache_rya_indexing_mongo_MongoEntityIndexTest/'><strong>org.apache.rya.indexing.mongo.MongoEntityIndexTest.org.apache.rya.indexing.mongo.MongoEntityIndexTest</strong></a></li></ul>

File: common/rya.api/pom.xml [hunk 1]
```diff
@@ -95,5 +95,20 @@ under the License.
...
+    <build>
+        <plugins>
+            <plugin>
+                <groupId>org.apache.rat</groupId>
+                <artifactId>apache-rat-plugin</artifactId>
+                <configuration>
+                    <excludes>
+                        <!-- Trivial listing of classes to be loaded via 
+                            SPI -->
+                        <exclude>src/main/resources/META-INF/services/org.openrdf.query.algebra.evaluation.function.Function</exclude>
```
Comment (jdasch) [inline common/rya.api/pom.xml]: Remove this.  Add the license to the service loader.  # is the comment character
http://docs.oracle.com/javase/7/docs/api/java/util/ServiceLoader.html

Comment (kchilton2) [inline common/rya.api/pom.xml]: This file needs a license header.

File: common/rya.api/src/main/java/org/apache/rya/api/functions/DateTimeWithinPeriod.java [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+import org.openrdf.model.datatypes.XMLDatatypeUtil;
+import org.openrdf.model.vocabulary.FN;
+import org.openrdf.model.vocabulary.XMLSchema;
+import org.openrdf.query.algebra.evaluation.ValueExprEvaluationException;
+import org.openrdf.query.algebra.evaluation.function.Function;
+
+import com.google.common.base.Preconditions;
+
+/**
+ * This {@link Function} determines whether two datetimes occur within a specified period of time of one another. The
```
Comment (jdasch) [inline common/rya.api/src/main/java/org/apache/rya/api/functions/DateTimeWithinPeriod.java]: change to XMLSchema.DATETIME

File: common/rya.api/src/main/java/org/apache/rya/api/functions/DateTimeWithinPeriod.java [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+     * </ul>
+     * 
+     * @param valueFactory - factory for creating values
+     * @param values - array of Value arguments for this Function.
+     */
+    @Override
+    public Value evaluate(ValueFactory valueFactory, Value... values) throws ValueExprEvaluationException {
+
+        // general validation of input
+        Preconditions.checkArgument(values.length == 4);
```
Comment (jdasch) [inline common/rya.api/src/main/java/org/apache/rya/api/functions/DateTimeWithinPeriod.java]: Do you want these to throw a ValueExprEvaluationException in order to follow the Function interface contract?

File: common/rya.api/src/main/java/org/apache/rya/api/functions/OWLTime.java [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+    /**
+     * Months class of type DurationDescription in OWL-Time ontology
+     */
+    public static final URI MONTHS_URI;
+    /**
+     * Years class of type DurationDescription in OWL-Time ontology
+     */
+    public static final URI YEARS_URI;
+
+    public static boolean isValidDurationType(URI duration) {
```
Comment (kchilton2) [inline common/rya.api/src/main/java/org/apache/rya/api/functions/OWLTime.java]: Missing javadocs.

Comment (kchilton2) [reply common/rya.api/src/main/java/org/apache/rya/api/functions/OWLTime.java]: Also, null check?

File: common/rya.api/src/main/java/org/apache/rya/api/functions/OWLTime.java [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+            Duration durEnum = getDurationFromURI(uri);
+            return duration * durEnum.millisConversion();
+        }
+
+        /**
+         * Provides the Duration from the given duration URI.
+         * @param uri
+         * @return - Duration enum type
+         */
+        public static Duration getDurationFromURI(URI uri) {
```
Comment (kchilton2) [inline common/rya.api/src/main/java/org/apache/rya/api/functions/OWLTime.java]: This could be map based. Build out a static map of URI to Duration, then return it from here.

File: common/rya.api/src/main/java/org/apache/rya/api/functions/OWLTime.java [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+     * Years class of type DurationDescription in OWL-Time ontology
+     */
+    public static final URI YEARS_URI;
+
+    public static boolean isValidDurationType(URI duration) {
+        return (duration.equals(SECONDS_URI) || duration.equals(MINUTES_URI) || duration.equals(HOURS_URI) || duration.equals(DAYS_URI)
+                || duration.equals(WEEKS_URI) || duration.equals(MONTHS_URI) || duration.equals(YEARS_URI));
+    }
+
+    static {
```
Comment (kchilton2) [inline common/rya.api/src/main/java/org/apache/rya/api/functions/OWLTime.java]: It would be nice if you didn't have this static method or the static fields up top.

You could have a private static final ValueFatory that you use within the Duration enum to build the URIs.

File: common/rya.api/src/main/java/org/apache/rya/api/functions/OWLTime.java [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+        SECONDS_URI = factory.createURI(NAMESPACE, "seconds");
+        MINUTES_URI = factory.createURI(NAMESPACE, "minutes");
+        HOURS_URI = factory.createURI(NAMESPACE, "hours");
+        DAYS_URI = factory.createURI(NAMESPACE, "days");
+        WEEKS_URI = factory.createURI(NAMESPACE, "weeks");
+        MONTHS_URI = factory.createURI(NAMESPACE, "months");
+        YEARS_URI = factory.createURI(NAMESPACE, "years");
+    }
+
+    public static enum Duration {
```
Comment (kchilton2) [inline common/rya.api/src/main/java/org/apache/rya/api/functions/OWLTime.java]: This class could use some javadocs.

File: common/rya.api/src/main/java/org/apache/rya/api/functions/DateTimeWithinPeriod.java [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+ */
+public class DateTimeWithinPeriod implements Function {
+
+    private static final String DATE_TIME_WITHIN = "dateTimeWithin";
+
+    @Override
+    public String getURI() {
+        return FN.NAMESPACE.toString() + DATE_TIME_WITHIN;
+    }
+
```
Comment (jdasch) [inline common/rya.api/src/main/java/org/apache/rya/api/functions/DateTimeWithinPeriod.java]: Change to 
```
    private static final String FUNCTION_URI = FN.NAMESPACE + "dateTimeWithin";

    @Override
    public String getURI() {
        return FUNCTION_URI;
    }
```

File: common/rya.api/src/main/java/org/apache/rya/api/functions/OWLTime.java (hunk 4 above)
Comment (kchilton2) [reply common/rya.api/src/main/java/org/apache/rya/api/functions/OWLTime.java]: This method could also be on the Duration enumeration. Or just get and if it doesn't find a value for the provided URI, then it is implicitly not a valid Duration type.

File: common/rya.api/src/main/java/org/apache/rya/api/functions/OWLTime.java (hunk 5 above)
Comment (kchilton2) [inline common/rya.api/src/main/java/org/apache/rya/api/functions/OWLTime.java]: This could also return an Optional<Duration> if you would rather indicate it is invalid by showing no value exists for the URI instead of throwing an exception.

File: common/rya.api/src/test/java/org/apache/rya/api/functions/DateTimeWithinTest.java [hunk 9]
```diff
@@ -0,0 +1,184 @@
...
+
+    private static final ValueFactory vf = new ValueFactoryImpl();
+    private static final Literal TRUE = vf.createLiteral(true);
+    private static final Literal FALSE = vf.createLiteral(false);
+
+    @Test 
+    public void testSeconds() throws DatatypeConfigurationException, ValueExprEvaluationException {
+        DatatypeFactory dtf = DatatypeFactory.newInstance();
+        
+        ZonedDateTime zTime = ZonedDateTime.now();
```
Comment (kchilton2) [inline common/rya.api/src/test/java/org/apache/rya/api/functions/DateTimeWithinTest.java]: This test is a little hard to follow. Maybe call this variable nowZTime?

File: common/rya.api/src/test/java/org/apache/rya/api/functions/DateTimeWithinTest.java [hunk 10]
```diff
@@ -0,0 +1,184 @@
...
+        String time = zTime.format(DateTimeFormatter.ISO_INSTANT);
+
+        ZonedDateTime zTime1 = zTime.minusSeconds(1);
```
Comment (kchilton2) [inline common/rya.api/src/test/java/org/apache/rya/api/functions/DateTimeWithinTest.java]: Maybe call this oneSecondEarlierZTime?

File: common/rya.api/src/main/java/org/apache/rya/api/functions/DateTimeWithinPeriod.java [hunk 11]
```diff
@@ -0,0 +1,121 @@
...
+ * can be found in the class {@link OWLTime} and below
+ * <ul>
+ * <li>http://www.w3.org/2006/time#days</li>
+ * <li>http://www.w3.org/2006/time#hours</li>
+ * <li>http://www.w3.org/2006/time#minutes</li>
+ * <li>http://www.w3.org/2006/time#seconds</li>
+ * </ul>
+ * 
+ */
+public class DateTimeWithinPeriod implements Function {
```
Comment (jdasch) [inline common/rya.api/src/main/java/org/apache/rya/api/functions/DateTimeWithinPeriod.java]: Rename `DateTimeWithinPeriod` to `DateTimeWithin`
or
Rename `DateTimeWithinTest` to `DateTimeWithinPeriodTest`

File: common/rya.api/src/main/java/org/apache/rya/api/functions/DateTimeWithinPeriod.java [hunk 12]
```diff
@@ -0,0 +1,121 @@
...
+     * 
+     * @param valueFactory - factory for creating values
+     * @param values - array of Value arguments for this Function.
+     */
+    @Override
+    public Value evaluate(ValueFactory valueFactory, Value... values) throws ValueExprEvaluationException {
+
+        // general validation of input
+        Preconditions.checkArgument(values.length == 4);
+        Preconditions.checkArgument(values[0] instanceof Literal);
```
Comment (kchilton2) [inline common/rya.api/src/main/java/org/apache/rya/api/functions/DateTimeWithinPeriod.java]: This is a style comment. It would be a little easier to understand that this code is doing if these array indices were descriptive of what was stored in the array at that position.

for exmple:
private static final int DATE_TIME_1 = 0;
private static final int DATE_TIME_2 = 1;
private static final int PERIOD = 2;
private static final int PERIOD_UNITS = 3;

Then the rest of the code could look like:
long epochTime1 = convertDatetimeToMillis((Literal) values[DATE_TIME_1]);
long epochTime2 = convertDatetimeToMillis((Literal) values[DATE_TIME_2]);
long periodMillis = convertPeriodToMillis((Literal) values[PERIOD], (URI) values[PERIOD_UNITS]);

etc

File: common/rya.api/src/main/java/org/apache/rya/api/functions/OWLTime.java [hunk 13]
```diff
@@ -0,0 +1,147 @@
...
+        SECONDS(1000, SECONDS_URI), 
+        MINUTES(60000, MINUTES_URI), 
+        HOURS(3600000, HOURS_URI), 
+        DAYS(86400000, DAYS_URI), 
+        WEEKS(604800000, WEEKS_URI), 
+        MONTHS(2592000000L, MONTHS_URI), 
+        YEARS(31536000000L, YEARS_URI);
+
+        private final URI uri;
+        private final long millisConversion;
```
Comment (kchilton2) [inline common/rya.api/src/main/java/org/apache/rya/api/functions/OWLTime.java]: Checkout TimeUnits.

File: common/rya.api/src/main/java/org/apache/rya/api/functions/DateTimeWithinPeriod.java (hunk 12 above)
Comment (meiercaleb) [reply common/rya.api/src/main/java/org/apache/rya/api/functions/DateTimeWithinPeriod.java]: Point taken.  I think the comments clearly explain the order of the arguments.

File: common/rya.api/src/test/java/org/apache/rya/api/functions/DateTimeWithinTest.java (hunk 9 above)
Comment (meiercaleb) [reply common/rya.api/src/test/java/org/apache/rya/api/functions/DateTimeWithinTest.java]: Changed names.

File: common/rya.api/src/test/java/org/apache/rya/api/functions/DateTimeWithinTest.java (hunk 10 above)
Comment (meiercaleb) [reply common/rya.api/src/test/java/org/apache/rya/api/functions/DateTimeWithinTest.java]: Updated names.

File: common/rya.api/src/main/java/org/apache/rya/api/functions/OWLTime.java (hunk 4 above)
Comment (meiercaleb) [reply common/rya.api/src/main/java/org/apache/rya/api/functions/OWLTime.java]: Done.

File: common/rya.api/src/main/java/org/apache/rya/api/functions/OWLTime.java (hunk 6 above)
Comment (meiercaleb) [reply common/rya.api/src/main/java/org/apache/rya/api/functions/OWLTime.java]: Done.

File: common/rya.api/src/main/java/org/apache/rya/api/functions/OWLTime.java (hunk 7 above)
Comment (meiercaleb) [reply common/rya.api/src/main/java/org/apache/rya/api/functions/OWLTime.java]: This class has been removed.

File: common/rya.api/src/main/java/org/apache/rya/api/functions/OWLTime.java (hunk 5 above)
Comment (meiercaleb) [reply common/rya.api/src/main/java/org/apache/rya/api/functions/OWLTime.java]: Done.

Comment (meiercaleb) [reply common/rya.api/src/main/java/org/apache/rya/api/functions/OWLTime.java]: This method has been removed.

File: common/rya.api/src/main/java/org/apache/rya/api/functions/OWLTime.java (hunk 13 above)
Comment (meiercaleb) [reply common/rya.api/src/main/java/org/apache/rya/api/functions/OWLTime.java]: Using ChronoUnit now.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/incubator-rya-master-with-optionals-pull-requests/517/

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/incubator-rya-master-with-optionals-pull-requests/518/

File: common/rya.api/src/main/java/org/apache/rya/api/functions/DateTimeWithinPeriod.java [hunk 14]
```diff
@@ -0,0 +1,128 @@
...
+     * <li>http://www.w3.org/2006/time#hours</li>
+     * <li>http://www.w3.org/2006/time#minutes</li>
+     * <li>http://www.w3.org/2006/time#seconds</li>
+     * </ul>
+     *
+     * @param valueFactory - factory for creating values
+     * @param values - array of Value arguments for this Function.
+     */
+    @Override
+    public Value evaluate(ValueFactory valueFactory, Value... values) throws ValueExprEvaluationException {
```
Comment (kchilton2) [inline common/rya.api/src/main/java/org/apache/rya/api/functions/DateTimeWithinPeriod.java]: Add null checks.

File: common/rya.api/src/main/java/org/apache/rya/api/functions/OWLTime.java [hunk 15]
```diff
@@ -0,0 +1,114 @@
...
+        return duration * unit.get().getDuration().toMillis();
+    }
+
+    /**
+     * Converts the {@link OWLTime} URI time unit to a {@link ChronoUnit} time unit
+     *
+     * @param durationURI - OWLTime time unit URI
+     * @return - corresponding ChronoUnit time unit
+     */
+    public static Optional<ChronoUnit> getChronoUnitFromURI(URI durationURI) {
```
Comment (kchilton2) [inline common/rya.api/src/main/java/org/apache/rya/api/functions/OWLTime.java]: This function could be simplified to:
```java
public static Optional<ChronoUnit> getChronoUnitFromURI(URI durationURI) {
    return Optional.ofNullable(DURATION_MAP.get(durationURI));
}
```

File: common/rya.api/src/main/java/org/apache/rya/api/functions/OWLTime.java [hunk 16]
```diff
@@ -0,0 +1,114 @@
...
+        DURATION_MAP.put(SECONDS_URI, ChronoUnit.SECONDS);
+        DURATION_MAP.put(MINUTES_URI, ChronoUnit.MINUTES);
+        DURATION_MAP.put(HOURS_URI, ChronoUnit.HOURS);
+        DURATION_MAP.put(DAYS_URI, ChronoUnit.DAYS);
+        DURATION_MAP.put(WEEKS_URI, ChronoUnit.WEEKS);
+    }
+
+    /**
+     * Verifies whether URI is a valid OWL-Time URI that is supported by this class.
+     * @param durationURI - OWLTime URI indicating the time unit
```
Comment (kchilton2) [inline common/rya.api/src/main/java/org/apache/rya/api/functions/OWLTime.java]: Document your nullness contract. 

```
@param durationURI - OWLTime URI indicating the time unit. (not null)
```

Comment (kchilton2) [reply common/rya.api/src/main/java/org/apache/rya/api/functions/OWLTime.java]: This is a general comment through out the code in this review.

File: common/rya.api/src/main/java/org/apache/rya/api/functions/OWLTime.java (hunk 15 above)
Comment (meiercaleb) [reply common/rya.api/src/main/java/org/apache/rya/api/functions/OWLTime.java]: Done.

File: common/rya.api/src/main/java/org/apache/rya/api/functions/DateTimeWithinPeriod.java (hunk 14 above)
Comment (meiercaleb) [reply common/rya.api/src/main/java/org/apache/rya/api/functions/DateTimeWithinPeriod.java]: Done.

File: common/rya.api/src/main/java/org/apache/rya/api/functions/OWLTime.java (hunk 16 above)
Comment (meiercaleb) [reply common/rya.api/src/main/java/org/apache/rya/api/functions/OWLTime.java]: Done.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/incubator-rya-master-with-optionals-pull-requests/519/

File: common/rya.api/src/main/java/org/apache/rya/api/functions/OWLTime.java [hunk 17]
```diff
@@ -0,0 +1,111 @@
...
+    /**
+     * Days class of type DurationDescription in OWL-Time ontology
+     */
+    public static final URI DAYS_URI = FACTORY.createURI(NAMESPACE, "days");
+    /**
+     * Weeks class of type DurationDescription in OWL-Time ontology
+     */
+    public static final URI WEEKS_URI = FACTORY.createURI(NAMESPACE, "weeks");
+
+    private static final Map<URI, ChronoUnit> DURATION_MAP = new HashMap<>();
```
Comment (jdasch) [inline common/rya.api/src/main/java/org/apache/rya/api/functions/OWLTime.java]: Nit: Using an ImmutableMap would make your intent here clearer.

File: common/rya.api/src/main/java/org/apache/rya/api/functions/DateTimeWithinPeriod.java [hunk 18]
```diff
@@ -0,0 +1,130 @@
...
+ * <li>http://www.w3.org/2006/time#days</li>
+ * <li>http://www.w3.org/2006/time#hours</li>
+ * <li>http://www.w3.org/2006/time#minutes</li>
+ * <li>http://www.w3.org/2006/time#seconds</li>
+ * </ul>
+ *
+ */
+public class DateTimeWithinPeriod implements Function {
+
+    private static final String FUNCTION_URI = FN.NAMESPACE + "dateTimeWithin";
```
Comment (jdasch) [inline common/rya.api/src/main/java/org/apache/rya/api/functions/DateTimeWithinPeriod.java]: Nit:  dateTimeWithinPeriod

File: common/rya.api/src/main/java/org/apache/rya/api/functions/OWLTime.java [hunk 19]
```diff
@@ -0,0 +1,111 @@
...
+        DURATION_MAP.put(WEEKS_URI, ChronoUnit.WEEKS);
+    }
+
+    /**
+     * Verifies whether URI is a valid OWL-Time URI that is supported by this class.
+     * @param durationURI - OWLTime URI indicating the time unit (not null)
+     * @return - {@code true} if this URI indicates a supported OWLTime time unit
+     */
+    public static boolean isValidDurationType(URI durationURI) {
+        checkNotNull(durationURI);
```
Comment (jdasch) [inline common/rya.api/src/main/java/org/apache/rya/api/functions/OWLTime.java]: Nit:  add user feedback message in the checkNotNull

File: common/rya.api/src/test/java/org/apache/rya/api/functions/DateTimeWithinPeriodTest.java [hunk 20]
```diff
@@ -0,0 +1,180 @@
...
+        Literal nowAustraliaTZ = vf.createLiteral(dtf.newXMLGregorianCalendar(time2));
+        Literal nowAsiaTZMinusOne = vf.createLiteral(dtf.newXMLGregorianCalendar(time3));
+
+        DateTimeWithinPeriod func = new DateTimeWithinPeriod();
+
+        assertEquals(TRUE, func.evaluate(vf, nowLocal, nowEuropeTZ, vf.createLiteral(1), OWLTime.SECONDS_URI));
+        assertEquals(TRUE, func.evaluate(vf, nowLocal, nowAustraliaTZ, vf.createLiteral(1), OWLTime.SECONDS_URI));
+        assertEquals(FALSE, func.evaluate(vf, nowLocal, nowAsiaTZMinusOne, vf.createLiteral(1), OWLTime.DAYS_URI));
+        assertEquals(TRUE, func.evaluate(vf, nowLocal, nowAsiaTZMinusOne, vf.createLiteral(2), OWLTime.DAYS_URI));
+    }
```
Comment (jdasch) [inline common/rya.api/src/test/java/org/apache/rya/api/functions/DateTimeWithinPeriodTest.java]: Nit: This feels cleaner to me...
```
    @Test
    public void testTimeZone2() throws DatatypeConfigurationException, ValueExprEvaluationException {
        final DatatypeFactory dtf = DatatypeFactory.newInstance();
        final Literal tPlus5 = vf.createLiteral(dtf.newXMLGregorianCalendar("2017-10-10T00:00:00+05:00"));
        final Literal tPlus8 = vf.createLiteral(dtf.newXMLGregorianCalendar("2017-10-10T00:00:00+08:00"));

        final DateTimeWithinPeriod func = new DateTimeWithinPeriod();

        assertEquals(FALSE, func.evaluate(vf, tPlus5, tPlus8, vf.createLiteral(3*60), OWLTime.MINUTES_URI));
        assertEquals(FALSE, func.evaluate(vf, tPlus5, tPlus8, vf.createLiteral(3), OWLTime.HOURS_URI));
        assertEquals(TRUE, func.evaluate(vf, tPlus5, tPlus8, vf.createLiteral(3*60+1), OWLTime.MINUTES_URI));
        assertEquals(TRUE, func.evaluate(vf, tPlus5, tPlus8, vf.createLiteral(4), OWLTime.HOURS_URI));
        assertEquals(TRUE, func.evaluate(vf, tPlus5, tPlus8, vf.createLiteral(1), OWLTime.DAYS_URI));
    }
```

File: common/rya.api/src/main/java/org/apache/rya/api/functions/DateTimeWithinPeriod.java [hunk 21]
```diff
@@ -0,0 +1,130 @@
...
+import org.openrdf.query.algebra.evaluation.function.Function;
+
+/**
+ * This {@link Function} determines whether two {@link XMLSchema#DATETIME}s occur within a specified period of time of
+ * one another. The method {@link Function#evaluate(ValueFactory, Value...)} expects four values, where the first two
+ * values are the datetimes, the third value is an integer indicating the period, and the fourth value is a URI
+ * indicating the time unit of the period. The URI must be of Type DurationDescription in the OWL-Time ontology (see
+ * <a href ="https://www.w3.org/TR/owl-time/">https://www.w3.org/TR/owl-time/</a>). Examples of valid time unit URIs can
+ * be found in the class {@link OWLTime} and below
+ * <ul>
```
Comment (jdasch) [inline common/rya.api/src/main/java/org/apache/rya/api/functions/DateTimeWithinPeriod.java]: Nit: As a user, I'd love a 5th optional argument that allows me to specify exclusiveness or inclusiveness of the period.
